pallets/loan/src/loan_type.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+//! Module provides functionality for different loan types
+use super::*;
+use scale_info::TypeInfo;
+use sp_arithmetic::traits::Zero;
+
+/// different types of loans
+#[derive(Encode, Decode, Copy, Clone, PartialEq, TypeInfo)]
+#[cfg_attr(any(feature = "std", feature = "runtime-benchmarks"), derive(Debug))]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum LoanType<Rate, Amount> {
+	BulletLoan(BulletLoan<Rate, Amount>),
+	CreditLine(CreditLine<Rate, Amount>),
+	CreditLineWithMaturity(CreditLineWithMaturity<Rate, Amount>),
+}
+
+impl<Rate, Amount> LoanType<Rate, Amount>
+where
+	Rate: FixedPointNumber,
+	Amount: FixedPointNumber,
+{
+	pub(crate) fn maturity_date(&self) -> Option<u64> {
+		match self {
+			LoanType::BulletLoan(bl) => Some(bl.maturity_date),
+			LoanType::CreditLine(_) => None,
+			LoanType::CreditLineWithMaturity(clm) => Some(clm.maturity_date),
+		}
+	}
+
+	pub(crate) fn is_valid(&self, now: u64) -> bool {
+		match self {
+			LoanType::BulletLoan(bl) => bl.is_valid(now),
+			LoanType::CreditLine(cl) => cl.is_valid(),
+			LoanType::CreditLineWithMaturity(clm) => clm.is_valid(now),
```

Comment:
Maybe introduce a `LoanType` trait that boundles `is_valid`, `present_value` etc. together to an interface or are the arguments needed for the different loan-type to heterogenous for that?

Reply:
no, make it a trait doen't help us with different func definition and args we need for speicifc loan types. 
That IMO would be too much abstraction

Reply:
Yeah, reasonable.